### ISSUE TITLE
Emit classes with private constructors.

### ISSUE THREAD

**Issue by mprobst** [angular/clutz#229]
e.g.

```js
function Foo() {
  /** @type {Foo.Bar_} */
  this.bar;
}

/** @private */
Foo.Bar_ = function() {};
```

**Comment by lambertjamesd** [comment]
```
goog.provide('test.PrivateConstructor');

/** 
 * @private
 * @constructor
 */
test.PrivateConstructor = function() {}

/**
 * @constructor
 */
test.PrivateConstructor.ChildPublicConstructor = function() {};
```
This will generate incorrectly too. The problem seems to be handling private constructors incorrectly. Right now clutz treats the entire as private when its constructor is marked as private. It should only make only the constructor private.

> Constructors marked @private can only be instantiated by code in the same file and by their static and instance members.

https://developers.google.com/closure/compiler/docs/js-for-compiler#tags

Visibility on constructors will be released with TypeScript 2.0
https://github.com/Microsoft/TypeScript/issues/2341

So for now, the best solution would be just to ignore the private annotation on constructors

**Comment by lambertjamesd** [comment]
This commit fixes the issue
https://github.com/lucidsoftware/clutz/pull/1/commits/4384cff2a6e5324d17e1dc4329bdf3895baab010

**Comment by rkirov** [comment]
I think the original bug report was different that the issue @lambertjamesd is bringing up. However, I cant repro the original issue, so lets continue on the private constructor issue (changing the issue title).

There are three distinct TypeScript concepts that are overloaded in the symbol `test.PrivateConstructor` - a type, a constructor for that type and a namespace where other object can live.

We agree that the constructor should be skipped (or emitted as private in future TS). It is definitely a bug that the namespace gets skipped because of the @private annotation.

The type is the tricky one, where I think closure *does* allow you to `goog.require` it and use it in a type position, but our generation is greatly simplified by collapsing those instances to `PrivateType` and not having to emit anything. Do you have opinions on this one?

I think there is a way to fix the issue with the namespace, while still not emitting the constructor (nor type), by directly calling `walkInnerSymbols` but not `walk` in the case of private constructor. I would prefer that change because, it is more incremental and we do have a large number of internal users.

**Comment by lambertjamesd** [comment]
A namespace wont cut it. The following works in closure
```
goog.provide('test.Foo');

/**
 * @private
 * @constructor
 */
test.Foo = function() {}

test.Foo.factory = function () {
  return new test.Foo();
}

test.Foo.prototype.method = function () {
  // Stuff
}
```
Some other file

```
goog.require('test.Foo')

test.Foo.factory().method()

```
Not only can it be a type, but you can access all of its members as well. The only thing marking the constructor as private will do is disallow calling new on it `new test.Foo()` 

I think the right approach is to simply ignore the annotation for constructors until TypeScript 2.0.

**Comment by rkirov** [comment]
That's a fair point, let's emit the constructor.

I think in that case `PrivateType` emitting code will need to be changed too, otherwise `factory` will have return type `PrivateType` which is not very usable from TS.

Can you add the `factory` static method to the private_types test in your PR?

**Comment by lambertjamesd** [comment]
Here is the pull request
https://github.com/angular/clutz/pull/300

**Comment by rkirov** [comment]
merged #300
